src/stats/cdf.rs

```diff
@@ -0,0 +1,211 @@
+// Copyright 2016 Johannes Köster.
+// Licensed under the MIT license (http://opensource.org/licenses/MIT)
+// This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Support for discrete probability distributions in terms of cumulative distribution
+//! functions (CDF).
+
+use std::f64;
+use std::iter;
+use std::slice;
+
+use num::traits::{cast, NumCast};
+use itertools::Itertools;
+
+use stats::logprobs::{self, LogProb};
+
+
+/// Implementation of a cumulative distribution function.
+#[derive(Debug, Clone)]
+pub struct CDF<T: PartialOrd> {
+    inner: Vec<(T, LogProb)>
+}
+
+
+impl<T: PartialOrd> CDF<T> {
+    /// Create CDF from given probability mass function (PMF). The PMF may contain duplicate values
+    /// the probabilities of which are summed during generation of the CDF.
+    ///
+    /// # Arguments
+    ///
+    /// * `pmf` - the PMF as a vector of value/probability pairs
+    pub fn from_pmf(mut entries: Vec<(T, LogProb)>) -> Self {
+        entries.sort_by(|&(ref a, _), &(ref b, _)| a.partial_cmp(b).unwrap());
+        let mut inner: Vec<(T, LogProb)> = Vec::new();
+        for mut e in entries.into_iter() {
+            let p = logprobs::add(inner.last().map_or(f64::NEG_INFINITY, |e| e.1), e.1);
```

Comment:
A very valid question. There are a couple of reasons (however I do not claim that this can't be discussed):

* Keep interoperability with non-logprobs and make it visible that this is an addition in log-space.
* During computations, logprobs can e.g. leave log space temporarily. Using newtypes and overloaded operators can make this quite inconvenient. However, I certainly want to look into this in the future.
* People expect a multiplication when doing a `+` on log probabilties.
* Most of the time, you will use sum instead of add, because it is faster and more stable. Add is only a last resort. It should not be the syntactically most pleasant solution.